From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: zfcp: Fix reference counter for point-to-point port
References: bnc#643173,LTC#67542
Patch-mainline: Yes

Symptom:     When using an point-to-point connection for FCP and
             running the sequence "enable FCP device", "disable FCP
             device", "detach FCP device (e.g. vmcp det)", this
             results in a hanging kernel thread kslowcrw.  Changing
             CCW devices is not possible anymore when this has happened.
Problem:     The reference counter of the point-to-point port is
             wrong. Detaching the FCP device tries to delete all
             attached ports, but it waits forever for the port's
             reference counter to drop to 0.
Solution:    Call zfcp_port_put after adding the point-to-point port.
             zfcp_port_enqueue returns with a reference counter of 1
             and the caller has to call zfcp_port_put when it does no
             longer need to hold a reference to the port.

Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/scsi/zfcp_erp.c |    1 +
 1 file changed, 1 insertion(+)

--- a/drivers/s390/scsi/zfcp_erp.c
+++ b/drivers/s390/scsi/zfcp_erp.c
@@ -625,6 +625,7 @@ static void zfcp_erp_enqueue_ptp_port(st
 	if (IS_ERR(port)) /* error or port already attached */
 		return;
 	_zfcp_erp_port_reopen(port, 0, "ereptp1", NULL);
+	zfcp_port_put(port);
 }
 
 static int zfcp_erp_adapter_strat_fsf_xconf(struct zfcp_erp_action *erp_action)
